components/nimbus/src/stateful/gecko_prefs.rs

```diff
@@ -0,0 +1,328 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+use crate::{
+    enrollment::{EnrollmentStatus, ExperimentEnrollment},
+    error::Result,
+    json::PrefValue,
+    EnrolledExperiment, Experiment, NimbusError,
+};
+use serde_derive::{Deserialize, Serialize};
+use serde_json::Value;
+use std::cmp::Ordering;
+use std::collections::{HashMap, HashSet};
+use std::fmt::{Display, Formatter};
+use std::sync::{Arc, Mutex, MutexGuard};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Copy)]
+#[serde(rename_all = "lowercase")]
+pub enum PrefBranch {
+    Default,
+    User,
+}
+
+impl Display for PrefBranch {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            PrefBranch::Default => f.write_str("default"),
+            PrefBranch::User => f.write_str("user"),
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GeckoPref {
+    pub pref: String,
+    pub branch: PrefBranch,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct GeckoPrefState {
+    pub gecko_pref: GeckoPref,
+    pub gecko_value: Option<PrefValue>,
+    pub enrollment_value: Option<PrefValue>,
+    pub is_user_set: bool,
+}
+
+impl GeckoPrefState {
+    pub fn new(pref: &str, branch: Option<PrefBranch>) -> Self {
+        Self {
+            gecko_pref: GeckoPref {
+                pref: pref.into(),
+                branch: branch.unwrap_or(PrefBranch::Default),
+            },
+            gecko_value: None,
+            enrollment_value: None,
+            is_user_set: false,
+        }
+    }
+
+    pub fn with_gecko_value(mut self, value: PrefValue) -> Self {
+        self.gecko_value = Some(value);
+        self
+    }
+
+    pub fn with_enrollment_value(mut self, value: PrefValue) -> Self {
+        self.enrollment_value = Some(value);
+        self
+    }
+
+    pub fn set_by_user(mut self) -> Self {
+        self.is_user_set = true;
+        self
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Hash, Eq, PartialEq, Copy)]
+pub enum PrefUnenrollReason {
+    Changed,
+    FailedToSet,
+}
+
+pub type MapOfFeatureIdToPropertyNameToGeckoPrefState =
+    HashMap<String, HashMap<String, GeckoPrefState>>;
+
+pub fn create_feature_prop_pref_map(
+    list: Vec<(&str, &str, GeckoPrefState)>,
+) -> MapOfFeatureIdToPropertyNameToGeckoPrefState {
+    list.iter().fold(
+        HashMap::new(),
+        |mut feature_map, (feature_id, prop_name, pref_state)| {
+            let feature_id = feature_id.to_string();
+            let prop_name = prop_name.to_string();
+            feature_map
+                .entry(feature_id)
+                .or_default()
+                .insert(prop_name.clone(), pref_state.clone());
```

Comment:
```suggestion
        |mut feature_map, (feature_id, prop_name, pref_state)| {
            feature_map
                .entry(feature_id.to_string())
                .or_default()
                .insert(prop_name.to_string(), pref_state.clone());
```

you don't need to to_string() and clone for prop_name